api/cpp/include/slint_platform.h

```diff
@@ -90,25 +89,23 @@ class WindowAdapter : public AbstractWindowAdapter
         cbindgen_private::slint_window_adapter_new(
                 this, [](void *wa) { delete reinterpret_cast<const WA *>(wa); },
                 [](void *wa) {
-                    return reinterpret_cast<const WA *>(wa)->m_renderer.renderer_handle();
+                    return reinterpret_cast<const WA *>(wa)->m_renderer->renderer_handle();
                 },
                 [](void *wa) { reinterpret_cast<const WA *>(wa)->show(); },
                 [](void *wa) { reinterpret_cast<const WA *>(wa)->hide(); },
                 [](void *wa) { reinterpret_cast<const WA *>(wa)->request_redraw(); }, &self);
-        m_renderer.init(&self);
         was_initialized = true;
         return self;
     }
 
 public:
-    /// Construct a WindowAdapter.  The arguments are forwarded to initialize the renderer
-    template<typename... Args>
-    explicit WindowAdapter(Args... a) : m_renderer(std::forward<Args>(a)...)
-    {
-    }
+    /// Construct a WindowAdapter
+    explicit WindowAdapter() { }
+
+    void set_renderer(std::unique_ptr<R> renderer) { m_renderer = std::move(renderer); }
 
     /// Return a reference to the renderer that can be used to do the rendering.
-    const R &renderer() const { return m_renderer; }
+    const R &renderer() const { return *m_renderer; }
```

Comment:
The documentation shoud mention that it is UB to call this before set_renderer


Idea: make it pure virtual,  so set_renderer is not needed.

Reply:
Good idea!